subgraphs/osmosis/src/common/utils.ts

```diff
@@ -0,0 +1,221 @@
+import { BigInt, ethereum, BigDecimal, cosmos } from "@graphprotocol/graph-ts";
+import {
+  LiquidityPool as LiquidityPoolStore,
+  _TokenPrice,
+} from "../../generated/schema";
+import { MsgPoolParams } from "../modules/Decoder";
+import { PoolFeesType } from "./types";
+import {
+  getOrCreateToken,
+  getOrCreateDexAmmProtocol,
+  getOrCreateLiquidityPoolFee,
+} from "./initializer";
+import * as constants from "../common/constants";
+import * as utils from "../common/utils";
+
+export function bigDecimalToBigInt(n: BigDecimal): BigInt {
+  return BigInt.fromString(n.toString().split(".")[0]);
+}
+
+export function enumToPrefix(snake: string): string {
+  return snake.toLowerCase().replace("_", "-") + "-";
+}
+
+export function prefixID(enumString: string, ID: string): string {
+  return enumToPrefix(enumString) + ID;
+}
+
+export function readValue<T>(
+  callResult: ethereum.CallResult<T>,
+  defaultValue: T
+): T {
+  return callResult.reverted ? defaultValue : callResult.value;
+}
+
+export function calculateAverage(items: BigDecimal[]): BigDecimal {
+  let sum = BigDecimal.fromString("0");
+  for (let i = 0; i < items.length; i++) {
+    sum = sum.plus(items[i]);
+  }
+
+  return sum.div(
+    BigDecimal.fromString(BigInt.fromI32(items.length).toString())
+  );
+}
+
+export function updatePoolTVL(
+  liquidityPool: LiquidityPoolStore,
+  block: cosmos.HeaderOnlyBlock
+): bool {
+  const index = updateBaseTokenPrice(liquidityPool, block) as i32;
+  if (index < 0) {
+    return false;
+  }
+
+  const inputTokens = liquidityPool.inputTokens;
+  const token = getOrCreateToken(inputTokens[index]);
+  let lastPrice = constants.BIGDECIMAL_ZERO;
+  if (token.lastPriceUSD !== null) {
+    lastPrice = token.lastPriceUSD!;
+  }
+
+  const inputTokenBalances = liquidityPool.inputTokenBalances;
+  const inputTokenWeights = liquidityPool.inputTokenWeights;
+  const amountUSD = utils
+    .convertTokenToDecimal(inputTokenBalances[index], token.decimals)
+    .times(lastPrice);
+  liquidityPool.totalValueLockedUSD = amountUSD
+    .times(constants.BIGDECIMAL_HUNDRED)
+    .div(inputTokenWeights[index]);
+  liquidityPool.save();
+
+  return true;
+}
+
+function updateBaseTokenPrice(
+  liquidityPool: LiquidityPoolStore,
+  block: cosmos.HeaderOnlyBlock
+): number {
+  const inputTokens = liquidityPool.inputTokens;
+  let stableCoinIndex = -1;
+  let osmoIndex = -1;
+  let atomIndex = -1;
+  for (let i = 0; i < inputTokens.length; i++) {
```

Comment:
Is there any way all of this token pricing stuff can also be encapsulated in the token getter?